Update linter config

ggwpez opened the pull request:
- Fixes linter warnings
- Updates the linter config

Closes #132  
Depends on #256

matthiasgeihs commented on .golangci.yml:
```diff
@@ -9,95 +9,54 @@ run:
+linters-settings:
+  gocritic:
+    disabled-checks:
+      # Accept code in the form of object.Call(object), eg:
+      # assert.True(t, null.Equals(null), "Expected equality of zero address to itself")
```
Why would we want to allow that in general? This is an instance where I think it makes sense to disable the linter for a specific line.

ggwpez replied:
Ok i updated it.

matthiasgeihs commented on .golangci.yml:
```diff
@@ -9,95 +9,54 @@ run:
+  cyclop:
+    # 15 is used by go-report-card as well.
+    max-complexity: 15
+    skip-tests: true
```
Why skip tests?

ggwpez commented on .golangci.yml:
```diff
@@ -9,95 +9,54 @@ run:
+    - varnamelen # We often use single letter arguments.
+    - exhaustivestruct # We often have uninitialized fields of structs.
+    - errorlint # We have interface errors which an impl needs to return.
+    - nakedret # We use naked returns in long functions.
```
Ok i was under the impression that this always enforced named returns on functions, but it only enforces to use the return names when the function has named returns.  
It are 6 occurrences, i will fix them.

ggwpez replied:
Somehow this linter only reports it at the end of functions, anyway its fixed now.

ggwpez commented on .golangci.yml:
```diff
@@ -9,95 +9,54 @@ run:
+    - wsl # Formatting disabled for now.
+    - gochecknoinits # We use init functions.
+    - gci # We have our own import order
+    # Unimportant
```
Ups, i removed it.

ggwpez commented:
I addressed the comments. There are two points for discussion remaining:  
1. Do we want to forbid functions to start with `Get…`?
2. Enabling `revive` (formerly golint) reports:
```pre
exported: func name will be used as test.TestAddress by other packages, and that stutters; consider calling this Address
exported: func name will be used as test.TestAccountWithWalletAndBackend by other packages, and that stutters; consider calling this AccountWithWalletAndBackend
```
I think this is the reason why we normally named the generic tests with `Generic…Test`.  You can see the reported errors [here](https://github.com/hyperledger-labs/go-perun/runs/4225465309?check_suite_focus=true#step:4:73).
@matthiasgeihs

matthiasgeihs commented:
> 1. Do we want to forbid functions to start with `Get…`?

Typically we shouldn't use "Get...". I had a look and it seems like there are easy work arounds for the locations where we are using it. In many cases we can simply remove it.

> 2. Enabling `revive` (formerly golint) reports:
> 
> ```
> exported: func name will be used as test.TestAddress by other packages, and that stutters; consider calling this Address
> exported: func name will be used as test.TestAccountWithWalletAndBackend by other packages, and that stutters; consider calling this AccountWithWalletAndBackend
> ```
> 
> I think this is the reason why we normally named the generic tests with `Generic…Test`.

I see. However, I don't think "Generic...Test" is a good name (or idiomatic) either. I don't have the perfect solution for this. I see two options:
- Either we surpress the linter warning for the respective cases.
- We go back to the other naming schema.

I have a preference for 1, but as I said, it is also not perfect. Any other suggestions?

ggwpez commented:
> Typically we shouldn't use "Get...". I had a look and it seems like there are easy work arounds for the locations where we are using it. In many cases we can simply remove it.  

Yes i think so as well.

> I have a preference for 1, but as I said, it is also not perfect. Any other suggestions?

I dont mind, but i also dont have a better naming scheme :laughing: will suppress for now.

ggwpez commented:
All addressed. Changes are in the last [force push](https://github.com/hyperledger-labs/go-perun/compare/18effa18e75539a24d7ed91751af08e24c302668..d02fd58130ec8e98601eb01e863d3cf195e5d79d).  
@matthiasgeihs

matthiasgeihs commented on wire/net/simple/listener_internal_test.go:
```diff
@@ -80,7 +80,7 @@ func TestListener_Accept(t *testing.T) {
 
 		test.AssertNotTerminates(t, timeout, func() {
-			l.Accept()
+			_, _ = l.Accept()
```
Do something with the return values or use `//nolint`?

ggwpez replied:
See discussion with the `_ =` above.

matthiasgeihs replied:
See answer above.

ggwpez replied:
Already updated it in the last iteration.

ggwpez commented on apps/payment/app_internal_test.go:
```diff
@@ -38,9 +38,9 @@ func TestApp_ValidInit(t *testing.T) {
 
 	nildata := &channel.State{Data: nil}
-	assert.Panics(func() { app.ValidInit(nil, nildata) })
+	assert.Panics(func() { _ = app.ValidInit(nil, nildata) })
```
We talked about the use of `_ =` vs `// nolint` on monday with the example of  
`defer func() { _ = closer.Closer() } ()`  
as opposed to
`defer closer.Closer() // nolint: errcheck`  
and came to the conclusion that suppressing was worse.  
Which is why i now always used `_ = ` instead of suppressing the linter.

matthiasgeihs replied:
We talked about it and that was not my conclusion. You may have gotten me wrong. Next time ask if you are unsure and I will try to be more clear.

ggwpez replied:
Okay anyway, i updated it.  
@matthiasgeihs

ggwpez commented on backend/ethereum/channel/adjudicator_test.go:
```diff
@@ -36,6 +36,7 @@ import (
 const defaultTxTimeout = 2 * time.Second
 
 func testSignState(t *testing.T, accounts []*keystore.Account, state *channel.State) channel.Transaction {
+	t.Helper()
```
The doc of `t.Helper()` states this:  
```pre
Helper marks the calling function as a test helper function. When printing file and line information, that function will be skipped. Helper may be called simultaneously from multiple goroutines.
```
and was done to satisfy the `thelper` linter.

ggwpez commented on backend/ethereum/channel/backend.go:
```diff
@@ -226,7 +226,10 @@ func EncodeState(state *adjudicator.ChannelState) ([]byte, error) {
-		asset := part.(*Asset)
+		asset, ok := part.(*Asset)
+		if !ok {
+			log.Panic("wrong address type")
```
updated

ggwpez commented on backend/ethereum/channel/contractbackend.go:
```diff
@@ -33,8 +33,11 @@ import (
+	// How many blocks we query into the past for events.
+	startBlockOffset            = 100
+	contractBackendHeadBuffSize = 10
+)
```
moved it up

ggwpez commented on backend/sim/wallet/address.go:
```diff
@@ -68,33 +70,38 @@ func (a *Address) Bytes() []byte {
 
 // String converts this address to a human-readable string.
 func (a *Address) String() string {
+	const bytePerNibble = 4
```
The sim addresses have shorter encodings for easier debugging when reading the logs.  
Updated it to `length`.

matthiasgeihs replied:
I think you mean they have a shorter string representation. (The byte encoding still holds the full information.)

I am not convinced that cutting off a good chunk of the information for the string representation. In any case, this function could be a one-liner: `return fmt.Sprintf("0x%x...", a.X.Bytes()[:length])`.

ggwpez replied:
I put it in https://github.com/hyperledger-labs/go-perun/issues/261

ggwpez commented on channel/persistence/keyvalue/restorer.go:
```diff
@@ -92,7 +92,7 @@ func (pr *PersistRestorer) RestorePeer(addr wire.Address) (persistence.ChannelIt
 	}
 	itPeer := sortedkv.NewTable(pr.db, prefix.PeerDB+key).NewIterator()
-	defer itPeer.Close() // nolint: errcheck
+	defer itPeer.Close()
```
I dont know, seems that `golangci-lint` does not report this anymore. Their changelog is unreadable :shrug:

matthiasgeihs replied:
Probably that's why: https://golangci-lint.run/usage/configuration/

```
# EXC0001 errcheck: Almost all programs ignore errors on these functions and in most cases it's ok
- Error return value of .((os\.)?std(out|err)\..*|.*Close|.*Flush|os\.Remove(All)?|.*print(f|ln)?|os\.(Un)?Setenv). is not checked
```

ggwpez replied:
Ok interesting. Then i close the discussion.

ggwpez commented on channel/transaction.go:
```diff
@@ -61,9 +61,13 @@ func (t *Transaction) Decode(r io.Reader) error {
 		return errors.WithMessage(err, "decoding stateSet bit")
 	}
-	if (stateSet % 2) == 0 {
+	switch stateSet {
+	case 0:
 		t.State = nil
 		return nil
+	case 1:
+	default:
+		return errors.Errorf("unknown stateSet value: %v", stateSet)
 	}
```
I wanted to get rid of the magic number `2`, but then i saw that the code should better reject values different from `0` and `1` since they would decode incorrectly anyway.

ggwpez commented on channel/transaction_test.go:
```diff
@@ -26,7 +26,7 @@ import (
 	rng := pkgtest.Prng(t)
 	lengths := []int{2, 5, 10, 40}
-	var tests [][]bool
+	var tests [][]bool // nolint: prealloc
```
Since i dont know how to calculate the length, would be something like `len(lengths) * 4 + ?`.

matthiasgeihs replied:
Instead you can write `tests := [][]bool{}`.

ggwpez replied:
I thought this would cause another linter to fail, but it works.

ggwpez commented on client/proposal.go:
```diff
@@ -687,15 +692,15 @@ func (c *Client) enableVer1Cache() {
 	c.version1Cache.enabled--
 	for _, u := range c.version1Cache.cache {
-		go c.handleChannelUpdate(u.uh, u.p, u.m)
+		go c.handleChannelUpdate(ctx, u.uh, u.p, u.m)
```
This is done for the [contextcheck](https://github.com/sylvia7788/contextcheck) linter.

matthiasgeihs replied:
In some cases before we were using context Client.Ctx, and now you are using a different context there. I am not saying that it is necessarily the wrong context you are using. It is just really hard to follow through these changes within such a large PR. I suggest we postpone context linting to a separate PR.

ggwpez replied:
Factored the context changes out.

ggwpez commented on backend/ethereum/wallet/backend.go:
```diff
@@ -31,6 +32,10 @@ type Backend struct{}
 // ref https://github.com/ethereum/go-ethereum/blob/54b271a86dd748f3b0bcebeaf678dc34e0d6177a/crypto/signature_cgo.go#L66
 const SigLen = 65
 
+// SigVSubtract value that is subtracted from the last byte of a signature if
+// the last bytes exceeds it.
+const SigVSubtract = 27
```
It is now used in `backend/eth/chan/test/transactor.go sigFromRSV` function and would need to be public for that.

matthiasgeihs replied:
Usually I try to go by the following rule: Only expose what is useful to a user of the library. Don't expose things just for the sake of testing. If you need to access internals during testing, you can make it an internal test. Does that make sense?

In this specific case, I have a hard time imagining why a library user would need to access this constant. For testing, you could copy the variable to the test package. What do you think?

ggwpez replied:
I copied the variable and unexported it.

ggwpez commented on client/channelconn.go:
```diff
@@ -42,11 +42,11 @@ type channelConn struct {
 // subscribes on the subscriber to all messages regarding this channel.
-func newChannelConn(id channel.ID, peers []wire.Address, idx channel.Index, sub wire.Subscriber, pub wire.Publisher) (_ *channelConn, err error) {
+func newChannelConn(_ context.Context, id channel.ID, peers []wire.Address, idx channel.Index, sub wire.Subscriber, pub wire.Publisher) (_ *channelConn, err error) {
 	// relay to receive all update responses
 	relay := wire.NewRelay()
 	// we cache all responses for the lifetime of the relay
-	relay.Cache(context.Background(), func(*wire.Envelope) bool { return true })
+	relay.Cache(context.Background(), func(*wire.Envelope) bool { return true }) // nolint: contextcheck
```
The `contextcheck` linter wants to always have context passed down to a function when it uses one.  
In this case the context would be to short. I think `relay.Cache` should not accept a context anyway.

matthiasgeihs replied:
It seems like enabling the context check linter is more involved. Maybe we postpone it to a separate PR?

ggwpez replied:
Yes i moved it out.

ggwpez commented on client/client_persistence_test.go:
```diff
@@ -27,17 +27,18 @@ func TestPersistencePetraRobert(t *testing.T) {
 	defer cancel()
 
-	runTwoPartyTest(ctx, t, func(rng *rand.Rand) (setups []ctest.RoleSetup, roles [2]ctest.Executer) {
+	runTwoPartyTest(t, ctx, func(rng *rand.Rand) (setups []ctest.RoleSetup, roles [2]ctest.Executer) {
```
Was done for the `thelper` linter, but now it does not report an error if i change it back. Reverting.

ggwpez commented on client/restore.go:
```diff
@@ -24,21 +24,23 @@ import (
 )
 
-type channelFromSourceSig = func(*Client, *persistence.Channel, *Channel, ...wire.Address) (*Channel, error)
+type channelFromSourceSig = func(context.Context, *Client, *persistence.Channel, *Channel, ...wire.Address) (*Channel, error)
```
Seems that one of the functions that call `channelFromSourceSig` would otherwise not pass it down.

matthiasgeihs replied:
I suggest we postpone context linting to a separate PR.

ggwpez commented on client/restore_internal_test.go:
```diff
@@ -32,7 +34,11 @@ import (
 )
 
+// timeout for reconstructChannel calls.
+var reconstructTimeout = 10 * time.Second
```
From all the linters - no one catches this :facepalm: :hankey:

matthiasgeihs replied:
:D

ggwpez commented on client/sync.go:
```diff
@@ -71,7 +70,6 @@ func (c *Client) handleSyncMsg(peer wire.Address, msg *msgChannelSync) {
 func (c *Client) syncChannel(ctx context.Context, ch *persistence.Channel, p wire.Address) (err error) {
 	recv := wire.NewReceiver()
-	// nolint:errcheck
 	defer recv.Close() // ignore error
```
I dont know, `golangci-lint` does not report it anymore.

matthiasgeihs replied:
Seems common practice, hence they adopted the linter. See above more more details.

ggwpez commented on client/sync.go:
```diff
@@ -101,7 +99,10 @@ func (c *Client) syncChannel(ctx context.Context, ch *persistence.Channel, p wir
 		return errors.WithMessage(err, "receiving sync message")
 	}
-	msg := env.Msg.(*msgChannelSync) // safe by the predicate
+	msg, ok := env.Msg.(*msgChannelSync) // safe by the predicate
+	if !ok {
+		log.Panic("internal error: wrong message type")
+	}
```
We have a message filter predicate above, so the type should be correct already. I removed the comment.

matthiasgeihs replied:
Would be OK to use nolint here, but also OK to check again, I guess.

ggwpez commented on client/test/alice.go:
```diff
@@ -33,8 +33,9 @@ type Alice struct {
-func NewAlice(setup RoleSetup, t *testing.T) *Alice {
-	return &Alice{Proposer: *NewProposer(setup, t, 3)}
+func NewAlice(t *testing.T, setup RoleSetup) *Alice {
+	t.Helper()
```
For the `thelper` linter since it is not a test function itself but just a helper which accepts a `*testing.T`.

ggwpez commented on client/test/backend.go:
```diff
@@ -360,8 +360,8 @@ func (b *MockBackend) removeSubscription(ch channel.ID, sub *mockSubscription) {
 	i, ok := func() (int, bool) {
-		for i, sub_ := range b.eventSubs[ch] {
-			if sub == sub_ {
+		for i, s := range b.eventSubs[ch] {
```
The `revive` linter does not like underscores in names.

ggwpez commented on client/update.go:
```diff
@@ -209,7 +207,10 @@ func (c *Channel) updateGeneric(
 	}
 
-	acc := res.(*msgChannelUpdateAcc) // safe by predicate of the updateResRecv
+	acc, ok := res.(*msgChannelUpdateAcc) // safe by predicate of the updateResRecv
```
The predicate of `NewUpdateResRecv` already filters the message type, so there should only `msgChannelUpdateAcc` arrive here.

ggwpez commented on client/virtual_channel.go:
```diff
@@ -74,37 +77,38 @@ func (c *Channel) proposeVirtualChannelFunding(ctx context.Context, virtual *Cha
 	return err
 }
 
-const responseTimeout = 10 * time.Second          // How long we wait until the proposal response must be transmitted.
-const virtualFundingTimeout = 10 * time.Second    // How long we wait for a matching funding proposal.
-const virtualSettlementTimeout = 10 * time.Second // How long we wait for a matching settlement proposal.
```
Removed; it was unused.

ggwpez commented on log/log.go:
```diff
@@ -39,8 +39,8 @@ func Set(l Logger) {
 
-// Get returns the currently set framework logger.
-func Get() Logger {
+// Log returns the currently set framework logger.
+func Log() Logger {
```
Ok i dont mind.

ggwpez commented on log/log.go:
```diff
@@ -54,11 +54,9 @@ type StdLogger interface {
 	Printf(format string, args ...interface{})
 	Print(...interface{})
 	Println(...interface{})
```
The `StdLogger` has the same interface as the `stdlog.Log` struct. But that is quite the stretch if someone only looks at our types :laughing:

matthiasgeihs replied:
I was assuming something like that. Could you extend the function docu respectively?

ggwpez replied:
I dont understand why the linters dont fail here as well, since it are undocumented interface functions :nauseated_face:  
Created issue https://github.com/hyperledger-labs/go-perun/issues/262

ggwpez commented on wallet/sig.go:
```diff
@@ -82,18 +85,19 @@ func EncodeSparseSigs(w io.Writer, sigs []Sig) error {
-		for bitIdx := 0; bitIdx < 8 && sigIdx < len(*sigs); bitIdx, sigIdx = bitIdx+1, sigIdx+1 {
-			if ((mask[maskIdx] >> bitIdx) % 2) == 0 {
+		for bitIdx := 0; bitIdx < bitPerByte && sigIdx < len(*sigs); bitIdx, sigIdx = bitIdx+1, sigIdx+1 {
+			if ((mask[maskIdx] >> bitIdx) % two) == 0 {
```
Ok renamed to `binaryModulo`.

matthiasgeihs replied:
I think this is another instance where using `//nolint` would be justified. Anyways, also OK like this.

ggwpez commented on wallet/sig.go:
```diff
@@ -59,10 +61,11 @@ func EncodeSparseSigs(w io.Writer, sigs []Sig) error {
 	for i, sig := range sigs {
 		if sig != nil {
-			mask[i/8] |= 0x01 << (i % 8)
+			mask[i/bitPerByte] |= one << (i % bitPerByte)
```
Ok i applied a `nolint`.

ggwpez commented on wallet/test/address.go:
```diff
@@ -24,7 +24,7 @@ import (
 // TestAddress runs a test suite designed to test the general functionality of
 // an address implementation.
-func TestAddress(t *testing.T, s *Setup) {
+func TestAddress(t *testing.T, s *Setup) { // nolint: revive
```
I added
```pre
//nolint:revive
// func name will be used as test.TestAddress by other packages,
// and that stutters; consider calling this Address (revive).
```

matthiasgeihs replied:
I meant rather like `//nolint // explanation`. There is also a similar example here https://golangci-lint.run/usage/false-positives/.

Furthermore, `//nolint: revive` should apply only to the respective line, not to the whole function. You can write the following:

```go
func TestAccountWithWalletAndBackend(t *testing.T, s *Setup) { //nolint: revive // `test.Test...` stutters, but we accept that here.
```

ggwpez replied:
Weird, i tried that before as well but it did not work.

ggwpez commented on wallet/test/wallet.go:
```diff
@@ -43,7 +43,7 @@ type Setup struct {
 // a corresponding wallet and backend implementation.
 // This function should be called by every implementation of the wallet interface.
-func TestAccountWithWalletAndBackend(t *testing.T, s *Setup) {
+func TestAccountWithWalletAndBackend(t *testing.T, s *Setup) { // nolint: revive
```
Added
```pre
//nolint:revive
// func name will be used as test.TestAccountWithWalletAndBackend by other packages,
// and that stutters; consider calling this AccountWithWalletAndBackend (revive).
```

matthiasgeihs replied:
See https://github.com/hyperledger-labs/go-perun/pull/260#discussion_r752285107.

ggwpez replied:
updated

ggwpez commented on watcher/local/watcher.go:
```diff
@@ -100,6 +100,16 @@ type (
 	chInitializer func() (*ch, error)
 )
 
+// ErrSubChannelsPresent signals that an operation failed because subchannels
+// were present.
+var ErrSubChannelsPresent = stderrors.New("sub-channels present")
+
+// IsErrSubChannelsPresent returns whether the cause of the error was an
+// ErrSubChannelsPresent error.
+func IsErrSubChannelsPresent(err error) bool {
+	return errors.Is(err, ErrSubChannelsPresent)
+}
+
```
Using ad-hoc errors like `stderrors.New("sub-channels present")` is not to the liking of the `goerr113` linter.  
Sadly the value of `stderrors.New` is not a compile-time constant.

ggwpez commented on client/client.go:
```diff
@@ -178,11 +178,11 @@ func (c *Client) Handle(ph ProposalHandler, uh UpdateHandler) {
-			go c.handleChannelUpdate(uh, env.Sender, msg)
+			go c.handleChannelUpdate(context.TODO(), uh, env.Sender, msg)
 		case *virtualChannelFundingProposal:
-			go c.handleChannelUpdate(uh, env.Sender, msg)
+			go c.handleChannelUpdate(context.TODO(), uh, env.Sender, msg)
 		case *virtualChannelSettlementProposal:
-			go c.handleChannelUpdate(uh, env.Sender, msg)
+			go c.handleChannelUpdate(context.TODO(), uh, env.Sender, msg)
```
Oh right, this would only make a difference if the client gets closed while there is an on-going request, but i will revert it.

matthiasgeihs replied:
I suggest we postpone context linting to a separate PR.

ggwpez replied:
Yes i also did not like the ctx changes, will remove the liner for now. Otherwise it wants to have `nolint` at every call in the tree where a ctx could be passed, that is annoying as well.  
If we instead remove the ctx from the `relay.Cache` it would be fixed.

ggwpez commented:
It looks like i messed up some timeout, since a follow up branch for this one fails in the CI…

matthiasgeihs commented on backend/sim/wallet/address_internal_test.go:
```diff
@@ -45,7 +46,8 @@ func TestAddress_ByteArray(t *testing.T) {
 
 		addr := &Address{
 			X: new(big.Int).SetBytes(dest[:32]),
-			Y: new(big.Int).SetBytes(dest[32:])}
```
Why is the linter not complaining?

ggwpez commented:
> It looks like i messed up some timeout, since a follow up branch for this one fails in the CI…

Could not reproduce now after ~8 runs, maybe was false alarm with slow runners.  
The run in question was [this](https://github.com/perun-network/go-perun/actions/runs/1472566878/attempts/1).

matthiasgeihs commented:
> > It looks like i messed up some timeout, since a follow up branch for this one fails in the CI…
> 
> Could not reproduce now after ~8 runs, maybe was false alarm with slow runners. The run in question was [this](https://github.com/perun-network/go-perun/actions/runs/1472566878/attempts/1).

Might have been related to the context changes?

matthiasgeihs commented on backend/ethereum/channel/funder.go:
```diff
@@ -211,7 +213,7 @@ func (f *Funder) fundAssets(ctx context.Context, channelID channel.ID, req chann
 func (f *Funder) sendFundingTx(ctx context.Context, request channel.FundingReq, contract assetHolder, fundingID [32]byte) (txs []*types.Transaction, fatal error) {
 	bal := request.Agreement[contract.assetIndex][request.Idx]
-	// nolint: gocritic
+	//nolint:gocritic
```
This `//nolint` can be avoided by improving the structure:

```go
// sendFundingTx sends and returns the TXs that are needed to fulfill the
// funding request. It is idempotent.
func (f *Funder) sendFundingTx(ctx context.Context, request channel.FundingReq, contract assetHolder, fundingID [32]byte) (txs []*types.Transaction, fatal error) {
	bal := request.Agreement[contract.assetIndex][request.Idx]

	if bal == nil || bal.Sign() <= 0 {
		f.log.WithFields(log.Fields{"channel": request.Params.ID(), "idx": request.Idx}).Debug("Skipped zero funding.")
		return nil, nil
	}

	alreadyFunded, err := f.checkFunded(ctx, bal, contract, fundingID)
	if err != nil {
		return nil, errors.WithMessage(err, "checking funded")
	}

	if alreadyFunded {
		f.log.WithFields(log.Fields{"channel": request.Params.ID(), "idx": request.Idx}).Debug("Skipped second funding.")
		return nil, nil
	}

	return f.deposit(ctx, bal, wallet.Address(*contract.Address), fundingID)
}
```

ggwpez replied:
put it in https://github.com/hyperledger-labs/go-perun/pull/260/commits/cdc57f576b60eb39e266ad6c3200d500c519d2f3

matthiasgeihs reviewed (commented):
I couldn't check all the files again. I am assuming it is OK. Just a few open comments left.
